### PR TITLE
Support bucket refs in Encore.ts

### PR DESCRIPTION
**Pull request by eandre**

### CODE REVIEW

File: docs/ts/primitives/object-storage.md [hunk 1]
```diff
@@ -123,3 +123,49 @@ If an upload fails due to a precondition not being met (like if the object alrea
...
 
 Other errors are returned as `ObjectsError` errors (which the above errors also extend).
+
+## Using bucket references
+
+Encore uses static analysis to determine which services are accessing each bucket,
+and what operations each service is performing.
+
+That information is used to provision infrastructure correctly,
+render architecture diagrams, and configure IAM permissions.
```
Comment (marcuskohlberg) [inline docs/ts/primitives/object-storage.md:133]:
```suggestion
That information is used for features such as rendering architecture diagrams, and can be used by Encore's Cloud Platform to provision infrastructure correctly and configure IAM permissions.
```

File: docs/ts/primitives/object-storage.md [hunk 2]
```diff
@@ -123,3 +123,49 @@ If an upload fails due to a precondition not being met (like if the object alrea
...
+
+This means that `Bucket` objects can't be passed around however you'd like,
+as it makes static analysis impossible in many cases. To work around these restrictions
+Encore allows you to get a "reference" to a bucket that can be passed around any way you want.
```
Comment (marcuskohlberg) [inline docs/ts/primitives/object-storage.md:137]:
```suggestion
This means `Bucket` objects can't be passed around however you like,
as it makes static analysis impossible in many cases. To simplify your workflow, given these restrictions,
Encore supports defining a "reference" to a bucket that can be passed around any way you want.
```

File: docs/ts/primitives/object-storage.md [hunk 3]
```diff
@@ -123,3 +123,49 @@ If an upload fails due to a precondition not being met (like if the object alrea
 and the `notExists: true` option is set), it throws a `PreconditionFailed` error.
 
 Other errors are returned as `ObjectsError` errors (which the above errors also extend).
+
+## Using bucket references
```
Comment (marcuskohlberg) [inline docs/ts/primitives/object-storage.md:127]:
```suggestion
## Bucket references
```

File: docs/ts/primitives/object-storage.md [hunk 4]
```diff
@@ -123,3 +123,49 @@ If an upload fails due to a precondition not being met (like if the object alrea
...
+
+That information is used to provision infrastructure correctly,
+render architecture diagrams, and configure IAM permissions.
+
+This means that `Bucket` objects can't be passed around however you'd like,
+as it makes static analysis impossible in many cases. To work around these restrictions
+Encore allows you to get a "reference" to a bucket that can be passed around any way you want.
+
+This is done by calling `bucket.ref<DesiredPermissions>()`, where `DesiredPermissions` is one of the pre-defined
+permission types defined in the `encore.dev/storage/objects` module.
```
Comment (marcuskohlberg) [inline docs/ts/primitives/object-storage.md:140]:
```suggestion
### Using bucket references

Define a bucket reference by calling `bucket.ref<DesiredPermissions>()` from within a service, where `DesiredPermissions` is one of the pre-defined permission types defined in the `encore.dev/storage/objects` module.
```

File: tsparser/src/parser/resources/infra/objects.rs [hunk 5]
```diff
@@ -123,11 +144,106 @@ pub fn resolve_bucket_usage(data: &ResolveUsageData, bucket: Lrc<Bucket>) -> Res
...
+        if depth > 10 {
+            // Prevent infinite recursion.
+            return None;
+        }
+
+        match t {
+            Type::Named(named) => {
+                let ops = match named.obj.name.as_deref() {
+                    Some("Lister") => vec![Operation::ListObjects],
+                    Some("Exister") | Some("Attrser") => vec![Operation::GetObjectMetadata],
```
Comment (fredr) [inline tsparser/src/parser/resources/infra/objects.rs:168]: The `Exister` permissions here isn't referenced in the docs and in the go-code, is it in use somewhere else?

Comment (eandre) [reply tsparser/src/parser/resources/infra/objects.rs]: Good catch, fixed :)

**Review by fredr — DISMISSED** [review]
:ship:

**Review by fredr — APPROVED** [review]

**Pull request merged**